### DISCUSSION TITLE
why DecodeAll failed?

### DISCUSSION THREAD

**Discussion by realcp1018** [klauspost/compress#859] [Q&A] [answered]
For the following 2 functions, I can get a correct dst with the first one, while the 2nd one will have a zeroed dst. 
What is the proper way to use DecodeAll()?
```go
// I know the uncompressed size, so I will pre-define a dst(`make([]byte, <size>)`) before use zstdDeCompress
func zstdDeCompress(input, dst []byte) error {
	decoder, err := zstd.NewReader(bytes.NewBuffer(input))
	defer decoder.Close()
	if err != nil {
		return err
	}
	_, err = decoder.Read(dst)
	if err != nil {
		return err
	}
	return nil
}
```
```go
func zstdDeCompress(input, dst []byte) error {
	decoder, err := zstd.NewReader(nil)
	defer decoder.Close()
	if err != nil {
		return err
	}
	dstNew, err = decoder.DecodeAll(input, dst)
	if err != nil {
		return err
	}
	dst = dstNew
	return nil
}
```

**Comment by klauspost** [comment, marked as answer]
[DecodeAll](https://pkg.go.dev/github.com/klauspost/compress@v1.16.7/zstd#Decoder.DecodeAll) appends to dst.

Using `decoder.DecodeAll(input, dst[:0:len(dst)])` with and also having the [WithDecodeAllCapLimit](https://pkg.go.dev/github.com/klauspost/compress@v1.16.7/zstd#WithDecodeAllCapLimit) should do it.

Remember to re-use your decoder and not create a new one every time.

It does of course not check if dst is over-sized, so you will need to check that and return an error yourself.

**Reply by realcp1018** [reply]
Thanks for your answer, and yes I will create/close a decoder outside of func for reuse.
